application/test/controllers/dataverse/datasets_controller_test.rb

```diff
@@ -77,6 +103,32 @@ def files_incomplete_no_data_file_json
     assert_equal "Dataverse service error. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
   end
 
+  test "should redirect to internal referrer after raising exception" do
+    Dataverse::DatasetService.any_instance.stubs(:find_dataset_version_by_persistent_id).raises("error")
+    Dataverse::DatasetService.any_instance.stubs(:search_dataset_files_by_persistent_id).returns(nil)
+    internal_referer = view_dataverse_landing_path.to_s
+    get view_dataverse_dataset_url(@new_id, "random_id"), headers: { "HTTP_REFERER" => internal_referer }
+    assert_redirected_to internal_referer
+    assert_equal "Dataverse service error. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
+  end
+
+  test "should redirect to internal referrer after raising exception with script name" do
+    Dataverse::DatasetService.any_instance.stubs(:find_dataset_version_by_persistent_id).raises("error")
+    Dataverse::DatasetService.any_instance.stubs(:search_dataset_files_by_persistent_id).returns(nil)
+    internal_referer = "/pun/sys/loop" + view_dataverse_landing_path.to_s
+    get view_dataverse_dataset_url(@new_id, "random_id"), headers: { "HTTP_REFERER" => internal_referer }, env: { "SCRIPT_NAME" => "/pun/sys/loop" }
+    assert_redirected_to internal_referer
+    assert_equal "Dataverse service error. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
+  end
+
+  test "should redirect to fallback url after raising exception coming from external referrer" do
```

Comment:
fallback is an implementation details. This test asserts that redirects to the root path.